### ISSUE TITLE
How can mimalloc manage a piece of allocated memory?

### ISSUE THREAD

**Issue by CS0522** [microsoft/mimalloc#922]
I have already allocated a piece of big memory by `mmap`, the memory size (as `MEM_SIZE`) is large, so I want to use it as a memory pool, or a heap, so that I can use the memory efficiently. For example, every time I want a piece of memory (large but not larger than `MEM_SIZE`), I can just easily use a function to allocate and get a `void *`, and release this pointer later, and the pool/heap can collect this memory. An example:

```cpp
// allocate memory
void* mem_ptr = mmap(NULL, MEM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);

// create a heap or memory pool based on the allocated memory
// is there a function works as create_heap(...)?
auto my_heap = create_heap(mem_ptr, MEM_SIZE);

// when I need to get some memory
void* mem_for_var1 = alloc(my_heap, mem_size_for_var1);
auto var1 = new (mem_for_var1) float();

// use var1

// release
...
```

And if there exists any memory allocator or memory pool library for C++ that can work as above, it is appreciated if you could tell me and leave some examples or the documents of it.

**Comment by daanx** [comment]
Yes, mimalloc has [`mi_manage_os_memory`](https://microsoft.github.io/mimalloc/group__extended.html#ga4c6486a1fdcd7a423b5f25fe4be8e0cf) for this purpose:
```C
bool mi_manage_os_memory(void* start, size_t size, bool is_committed, bool is_large, bool is_zero, int numa_node);
```

This will add the memory area to the mimalloc arena's from which it will allocate its memory (use `numa_node=-1` for no NUMA preference). There is no need for a specific heap, just `mi_malloc` will work. It is best if the area is as large as possible such that mimalloc can manage it most effectively with regard to fragmentation.

Hope this helps,
  Daan


ps. There is also an extended api, [`mi_manage_os_memory_ex`](https://microsoft.github.io/mimalloc/group__extended.html#ga41ce8525d77bbb60f618fa1029994f6e)
to reserve "exclusive" memory that can only be used with heaps attached to such memory (`mi_heap_new_in_arena`) but I don't think that is what you are asking?  (this is used for example to implement compressed pointers in the Koka language using a 4GiB exclusive area for language allocations)

**Comment by CS0522** [comment]
Thanks very much for your help! It really does me a favor. mimalloc is very useful, and my project cannot be done without it.

**Comment by gulmezmerve** [comment]
@daanx I am a bit confused about how the mimalloc uses the process memory initially. As far as I know, when the process starts, it already reserves some memory area for the heap, and libc uses sbrk to extend. 

But I can track mimalloc down, I understand that it does mmap do manage the memory, right?

**Comment by gulmezmerve** [comment]
@CS0522 When you call the mmap, you should give some hint for alignment, otherwise, you will get some assertion. I found manually what is the requirement as '0x400000'

**Comment by MohammadKhalaji** [comment]
Hi @daanx 

Thank you for having responded in the issue and similar issues. 

I have a similar problem to the OP, and I can't get mimalloc to allocate from the mmapped file, even with `mi_reserve_os_memory`... 

Here's the code (I've removed sanity and error message checks for brevity)

```C
#define FILE_SIZE (16 * 1024 * 1024) // 16MB

fd = open("mmapped_file.txt", O_RDWR | O_CREAT, 0666);
char *file_in_memory = mmap(NULL, FILE_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_32BIT, fd, 0);
int reserveRet = mi_reserve_os_memory(FILE_SIZE, false, true);
int manageRet = mi_manage_os_memory(file_in_memory, FILE_SIZE, true, false, false, -1);

void *p = mi_malloc(1 * 1024 * 1024); // 1MB
void *q = mi_malloc(1 * 1024 * 1024); // 1MB
void *r = mi_malloc(1 * 1024 * 1024); // 1MB
void *s = mi_malloc(1 * 1024 * 1024); // 1MB

printf("base: %p\n", file_in_memory);
printf("p   : %p\n", p);
printf("q   : %p\n", q);
printf("r   : %p\n", r);
printf("s   : %p\n", s);

```


Outputs: 
```
base: 0x40000000
p   : 0x20000000200
q   : 0x20000140200
r   : 0x20000280200
s   : 0x20000400200
```


How do I make sure `mimalloc` is using my mmapped file?

I'll try your solution in #730 next, but I'm not sure where to find `MIMALLOC_RESERVE_OS_MEMORY` or `MIMALLOC_LIMIT_OS_ALLOC`
